File: src/runtime/environment.ts
export default class Environment {
  private variables: Map<string, any> = new Map();
  private parent: Environment | null;

  constructor(par: Environment | null) {
    this.variables = new Map();
    this.parent = par; 
  }

  public declareVar(name: string, value: any): void {
    if(this.variables.has(name)){
      throw new Error("Cannot redefine variable");
    }

    this.variables.set(name, value);
    return;
  } 

  public assignVar(name: string, value: any): void {
    const env = this.resolveEnv(name);

    env.variables.set(name, value);
    return;
  }
  
  public resolveEnv(name: string): Environment {
    if(this.variables.has(name))
      return this;

    if (!this.parent) {
      throw new Error(`${name} has not been defined anywhere in program`); 
    }

    return this.parent.resolveEnv(name);
  }

  public lookupVar(name: string): any {
    const env = this.resolveEnv(name);

    return env.variables.get(name);
  }
}

File: src/runtime/environment.cpp
#include <format>
#include <stdexcept>

#include <scriptum/runtime/environment.hpp>

namespace scriptum {

Environment::Environment(Environment* parent) : m_parent(parent) {}

void Environment::declareVar(const std::string& name, const Value& value)
{
    if (m_variables.contains(name))
    {
        throw std::runtime_error(std::format("Cannot redefine variable: {}", name));
    }
    m_variables[name] = value;
}

void Environment::assignVar(const std::string& name, const Value& value)
{
    Environment* env = resolveEnv(name);
    env->m_variables.at(name) = value;
}

Value Environment::lookupVar(const std::string& name) const
{
    const Environment* env = resolveEnv(name);
    return env->m_variables.at(name);
}

Environment* Environment::resolveEnv(const std::string& name) const
{
    if (m_variables.contains(name))
    {
        return const_cast<Environment*>(this);
    }

    if (m_parent == nullptr)
    {
        throw std::runtime_error(std::format("{}: is not defined anywhere in program", name));
    }

    return m_parent->resolveEnv(name);
}

} // namespace scriptum
